### ISSUE TITLE
Data loss in Brave browser

### ISSUE THREAD

**Issue by Brave-Matt** [brave/brave-browser#40375]
### Description

This issue is a continuation of a previous and ongoing [issue](https://github.com/brave/brave-browser/issues/33548) that was recently closed. Previous issue description:

> Please note that this issue has been ongoing for sometime now.
> 
> We have noticed a significant uptick in user reports concerning a random/unexplained loss of data in the browser. Based on these reports, it appears that the data in question can be anything stored locally/in the browser — from passwords, to bookmarks, to all browser profile data. This issue occurs across all [desktop] platforms.
> 
> We started tracking individual instances of this issue occurring in a document here: https://docs.google.com/spreadsheets/d/1Caau3LFaktSXv4vmecmTkQjr7XOtuaQK9G2ozol43_Q/edit?pli=1#gid=0
> 
> Please also note that the above document is not representative of all the reports received and that particular document is not quite up to date at this time. We do however have an ongoing Community thread where we have been collecting Community user reports, along with additional instructions for these users to follow and extra steps they can take to get us the data we need:
> https://community.brave.com/t/notice-unexplained-data-loss-issue/485498
- - - 
The purpose of this issue is to continue tracking users losing data (passwords, bookmarks, history, etc) that proceeded (and continue to persist) the [password decryption issue](https://github.com/brave/brave-core/pulls?q=is%3Apr+%22Uplift+of+%2325040%22+%22Enbl+%60SkipUndecryptablePasswords%60+on+Linux+%26+macOS%22+%22uplift+to%22+) that was introduced with the bump to Chromium 127.

The issue itself is largely unchanged — users are seeing different data types in the browser "disappear" due to know discernible reason/intentional action taken by the user. This includes any kind of data held in the browser, as shown in the sheet linked above. While I believe that most of these cases are _not_ due to user error/behavior, there are a few caveats where there are _known_ reasons in which this occurs.
### Known occurrences (will continue to add to this list) 
Scenarios in which data loss is expected include: 

**Keychain deleted or inaccessible**
For macOS users, the secret used to encrypt/decrypt is stored in the macOS Keychain. If this Keychain gets deleted, permission is lost (rendering the Keychain inaccessible), or if the [user's] profile folder is moved to a new machine (which doesn't have that keychain), then the new machine will not be able to read the secret and decrypt the data and data loss would be expected.

**Corrupted, deleted or moved `Local State` file**
For Windows users, the secret key is stored in the `Local State` file. If this file (or the entire profile folder) is copied and moved to a new machine, the new machine will be unable to decrypt the secret and a new key would be generated, resulting in data loss. 

Additionally, if the `Local State` file is corrupted (or if this file is deleted) this will likely result in a loss of data as well, as the secret will not be able to be decrypted and a new key will be created.

### Troubleshooting Steps/Helpful information
At this time, troubleshooting and _relevant_ information that users (or devs) can share on this is a bit scattered, so I will use this section to reorganize and highlight what information can help resolve this issue. For now, please refer to the "[Actions users can take](https://community.brave.com/t/notice-unexplained-data-loss-issue/485498#actions-users-can-take-to-help-4)" section in the Community thread for basic troubleshooting and information gathering.

**Additional troubleshooting:**
- **Finding `Local State.bad` file (Windows only):**
If your `Local State` file gets corrupted for any reason, it will move itself out of the way and change its name to `Local State.bad` and replace itself with a "fresh" `Local State` file (which, again, would present as a loss of data). You would be shown an error when launching into Brave:
![image](https://github.com/user-attachments/assets/09ed828d-ac07-4c7b-a38e-5896d8b6b65b)

You can look for this in your `User data` directory by going to `%LOCALAPPDATA%\BraveSoftware\Brave-Browser\User Data` on your PC. There, you should see the `Local State.bad` and the new `Local State`.

This error might happen if you edited the `Local State` file. It's in JSON format - you can try pasting the `Local State.bad` content into a JSON validator to ensure there isn't something wrong with the formatting.  The encryption key will be stored under the JSON object `os_crypt`. If you're up for some text editing, you can copy this node over from the `Local State.bad` into the new `Local State`.

### Steps to reproduce

N/A

### Actual result

N/A

### Expected result

Data stays safe and intact in browser 

### Reproduces how often

No steps to reproduce

### Brave version (brave://version info)

Issue has been present since v1.50x up to the most recent release. 

### Channel information

- [X] release (stable)
- [X] beta
- [X] nightly

### Reproducibility

- [ ] with Brave Shields disabled
- [ ] with Brave Rewards disabled
- [ ] in the latest version of Chrome

**Comment by szilardszaloki** [comment]
Row `15` [here](https://docs.google.com/spreadsheets/d/1Caau3LFaktSXv4vmecmTkQjr7XOtuaQK9G2ozol43_Q/edit?pli=1&gid=0#gid=0&range=A15:F15) — [issue](https://community.brave.com/t/randomly-lost-all-of-my-saved-passwords/484860) on Brave Community (reported by `mlchatnoir`):
- breaking commit in Brave: https://github.com/brave/brave-core/commit/f82ae5a3a6d581578519e30f37de7ef55da93108 (basically, `ReadPassword()` was falsely assumed to return a `bool`, whereas it returns a `KWalletDBus::Error` — and `!ReadPassword()` is not the same as `ReadPassword() != KWalletDBus::SUCCESS`, since `enum Error { SUCCESS = 0, CANNOT_CONTACT, CANNOT_READ };`)
- first went live with [1.51.106](https://github.com/brave/brave-browser/releases/tag/v1.51.106) on April 27, 2023 (`mlchatnoir` reported it on April 30, 2023)
- corresponding issue on GitHub: https://github.com/brave/brave-browser/issues/30147
- fixed via https://github.com/brave/brave-core/pull/18404 (uplifted into `1.51.x`)
- fix went live with [1.51.117](https://github.com/brave/brave-browser/releases/tag/v1.51.117) on May 15, 2023
- `mlchatnoir` also reported on May 18, 2023 that after an upgrade, passwords came back. ✅

**Comment by LaurenWags** [comment]
Per discussion via https://bravesoftware.slack.com/archives/CHGKGMHDJ/p1723569829908629?thread_ts=1723127438.506119&cid=CHGKGMHDJ, removed `OS/Windows` as the associated PR is only for macOS and Linux.

Added `release-notes/include` as well.

cc @rebron @kjozwiak

**Comment by bsclifton** [comment]
Issue to follow which help in some cases - filed upstream in Chromium:
https://issues.chromium.org/u/1/issues/365913981

**Comment by NonMeme** [comment]
I'm not sure if this is the right place to reply, but I am on Linux and Brave recently lost all saved passwords for me.

**Comment by hollons** [comment]
We've had an uptick of Password lost reports with 1.80.113 release. 7 in total. One of the users filed this issue https://github.com/brave/brave-browser/issues/47258.

**Comment by Saoiray** [comment]
Seeing recent influx of reports. Linux, Windows, etc.

* https://community.brave.com/t/all-passwords-deleted/631643
* https://community.brave.com/t/bug-report-brave-1-80-113-on-the-latest-version-deleted-all-saved-passwords-after-update/629274/
* https://community.brave.com/t/brave-forgot-all-my-sessions-and-passwords/630652/
* https://community.brave.com/t/all-passwords-disappeared/631429
* https://community.brave.com/t/brave-got-rid-of-all-my-passwords/631485
* https://community.brave.com/t/password-disapear/631424
* https://community.brave.com/t/passwords-stored-lost/631310
* https://community.brave.com/t/passwords-keep-being-deleted-and-all-sites-logged-out/630290
* https://community.brave.com/t/lost-all-saved-passwords-and-logged-out-upon-brave-closing-unexpectedly/629694
* https://community.brave.com/t/lost-all-my-password-data-how-do-i-restore-it-if-i-can/630268




Different but perhaps related:
* https://community.brave.com/t/new-users-removed/631541

### -NOTE-
Saw a post at https://community.brave.com/t/brave-does-not-save-my-login-info/623731 where person claims:


> I found it finally it’s a bug in the browser and I found out how to trigger it
>
> 1. open/reboot your desktop
> 2. press any link before opening brave for the first time
> 3. brave for some reason will delete all you’re login info
> 4.note : brave will not delete anything if you open brave before opening any link*
>
> it’s in the apt version of brave
>
> I hope this gets fixed soon


cc: @bsclifton, @hollons

**Comment by Saoiray** [comment]
@bsclifton just wanted to point out this seems to be an ongoing issue and have seen an recent influx of reports. These are appearing on Windows and Linux, so it doesn't seem isolated to any particular OS. Situations vary as well. For some it's when the browser crashes, other report it occurs when restarting, and yet others for things like when a browser update is released. 

Links below are just from Brave Community but there are more across Reddit and other sources as well. 

* https://community.brave.com/t/all-my-passwords-are-lost/636724 
* https://community.brave.com/t/brave-not-functioning-passwords-lost-and-brave-not-updating/636611
* https://community.brave.com/t/my-passwords-disappeared-from-brave/635916
* https://community.brave.com/t/suddenly-lost-all-my-saved-passwords-from-brave-but-did-nothing/633144
* https://community.brave.com/t/passwords-disappearing-w10-v-1-80-120-chromium-138-0-7204-101-official-build-64-bit/632075
* https://community.brave.com/t/passwords-keep-being-deleted-and-all-sites-logged-out/630290
* https://community.brave.com/t/lost-all-saved-passwords-and-autofill-data/613173
* https://community.brave.com/t/brave-removed-all-my-saved-passwords-and-logged-me-out-of-all-my-accounts/608347
* https://community.brave.com/t/everytime-i-try-to-get-back-the-unsaved-passwords-log-ins-they-dont-appear-in-the-saved-passwords/636869
* https://community.brave.com/t/being-logged-out-and-password-forgot-everyday-on-all-sites/636594
* https://community.brave.com/t/in-ubuntu-using-snap-after-an-update-passwords-are-gone-yes-i-had-them-backed-up/636080
* https://community.brave.com/t/saved-passwords-lost-all-of-the-sudden/635627
* https://community.brave.com/t/brave-signs-out-of-most-websites-and-deletes-passwords-after-restart-linux/621521
* https://community.brave.com/t/brave-browser-deletes-all-password-after-pc-restart/635075
* https://community.brave.com/t/brave-is-not-saving-passwords/635264
* https://community.brave.com/t/all-the-passwords-in-my-password-manager-just-disappeared/634641
* https://community.brave.com/t/all-passwords-are-gone-in-password-manager/623209
* https://community.brave.com/t/passwords-are-gone-need-any-sort-of-help-restoring-them/633388
* https://community.brave.com/t/brave-deletes-all-saved-passwords-after-every-pc-reboot/633329
* https://community.brave.com/t/brave-removed-all-my-saved-passwords-and-logs-me-out-of-all-sites-randomly/633239
* https://community.brave.com/t/bug-report-brave-1-80-113-on-the-latest-version-deleted-all-saved-passwords-after-update/629274
* https://community.brave.com/t/suddenly-lost-all-my-saved-passwords-from-brave-but-did-nothing/633144
* https://community.brave.com/t/all-passwords-deleted/631643
* https://community.brave.com/t/brave-got-rid-of-all-my-passwords/631485
* https://community.brave.com/t/all-passwords-disappeared/631429
* https://community.brave.com/t/password-disapear/631424
* https://community.brave.com/t/passwords-stored-lost/631310
* https://community.brave.com/t/lost-all-saved-passwords-and-logged-out-upon-brave-closing-unexpectedly/629694

**Comment by ourichermath** [comment]
I hope I'm doing the right thing by reporting this here.  Yesterday, for no reason at all, Brave lost 2 of my wife's 3 profiles.  One worked fine, but nothing would happen when attempting to launch an instance of the browser using the other two profiles.  Trying to edit or delete the profiles under Manage Brave Profiles would just crash that browser instance.

**Comment by LaurenWags** [comment]
Hi @ourichermath could you confirm which version and OS you are using? You can go to brave://version to grab the first 3 lines of that page which contain that info - Brave, Revision, OS.

Unfortunately the `1.83.108` release yesterday did have a crash on launch that occurred in certain cases. It's possible that's what your wife encountered. We're tracking that specific crash via https://github.com/brave/brave-browser/issues/49861. I can reproduce the crash from this issue that way (where some profiles work and some do not) as long as the initial profile created isn't affected.

There is a fix for https://github.com/brave/brave-browser/issues/49861 in testing currently. If all goes well, the fix will be available in [1.83.109](https://github.com/brave/brave-browser/releases/tag/v1.83.109) and released later today.

**Comment by ourichermath** [comment]
Sorry.  I should have thought about posting that information earlier.  Here it is:

Brave	1.83.108 Chromium: 141.0.7390.55 (Official Build) (64-bit) 
Revision	4c70823b107aeb49f74812a32744ce84f268d493
OS	Windows 11 Version 24H2 (Build 26100.6584)

**Comment by LaurenWags** [comment]
Thanks @ourichermath. We just released `1.83.109`, so update when you get a chance and if you can, let us know if that resolved your issue.

**Comment by ourichermath** [comment]
Updated to 1.83.109.  I had already manually recreated the deleted profiles, but I couldn't delete one of the old ones in Manage Brave Profiles under 1.83.108.  Under 1.83.109, that profile loaded up just fine.  So, it looks like the new version fixes my particular problem.  Thanks.

**Comment by YESNLars** [comment]
Just got it again (didn't update, not brave nor windows) i was gone for like a few hours, shutdown my pc, comeback, all logindata, search engines and profiles are gone (except when I start my profile shortcut it suddenly 'remembers' i have profiles (before there was no profile logo next to music and next to menu, and then it appears after i start another profile). It is very annoying to specifically add all my custom search engines again in all my profiles 😠 
I really hope this will get fixed, already got it 3 times earlier in this whole year. D:

oh and version:
v1.84.141

**Comment by af7567** [comment]
I had a few months without the problem occuring, but this week it went wrong for me too. I am on version 1.84.139 on Linux

**Comment by samcat74** [comment]
all passwords are just gone! WTF! first time for me in 5 years using Brave.

```
 $ brave-browser --version
Brave Browser 142.1.84.135
```

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 25.10
Release:        25.10
Codename:       questing
```

**Comment by Brave-Matt** [comment]
https://www.reddit.com/r/brave_browser/comments/1pdaxop/comment/ns8ftqf/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

https://community.brave.app/t/all-passwords-and-logins-gone/646505/3

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/my-profiles-have-vanished/646536

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-lost-all-passwords-settings-overnight/645592/4
https://community.brave.app/t/lost-all-cookies-including-passwords-and-site-data/643809/2
https://community.brave.app/t/suddenly-lost-all-passwords-again/641489/12
https://community.brave.app/t/passwords-lost-local-state-changed/640629
https://community.brave.app/t/i-lost-all-of-my-passwords/639368
https://community.brave.app/t/lost-all-my-bookmarks/644984
https://community.brave.app/t/update-deleted-everything-profile-bookmarks-extensions-settings/625142/25
https://community.brave.app/t/urgent-help-brave-deleted-all-my-bookmarks-after-an-upgrade-how-do-i-get-them-back/634247/10
https://community.brave.app/t/can-i-recover-all-lost-data-settings/645761
https://community.brave.app/t/brave-browser-suddenly-reset-lost-all-data-on-its-own/642404
https://community.brave.app/t/i-lost-all-my-data/639238/2
https://community.brave.app/t/brave-deleted-all-of-my-passwords/643457

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-lost-all-tabs-history-bookmarks-on-windows-11/646574/3

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/im-really-tired-of-having-all-my-profile-data-wiped-when-brave-does-an-update/646749

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/all-my-history-and-logins-are-wiped-after-latest-profile-update/646816/3

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/passwords-missing/646795

**Comment by Brave-Matt** [comment]
<img width="1036" height="269" alt="Image" src="https://github.com/user-attachments/assets/54d67181-8535-425a-a772-622ab994b5d8" />

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/everything-gone-after-update-passwords-bookmarks-top-sites/647130/2

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/browser-history-completely-empty-google-account-cookies-wiped/646948

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-sync-disconnects-after-uninstalling-powertoys/647212

**Comment by Brave-Matt** [comment]
User communicating via email encountering data loss:

> _OS: Win10 
Do not know the previous brave version current one: 1.85.116
I do not have Brave Sync enabled
What happened: Woke up opened my computer saw it was booting up (aka windows must have done some update). When I opened brave, I saw i had to pick a profile i want to use (this was new), tried all all 3 profiles ( default, default release, work) none of them had my browser history. Usually when i open up it offers me the option to restore, there was nothing. Then opened up history to restore it myself, well there was nothing to restore. Brave showed me the last page i had opened was from 28.03(march).2025 . Looked through the community forum one guy recommended me to go do c:/users/username/appdata/local/BraveSoftware/Brave-browser/User data/Default/ and look for Last_session, current_session etc. Did not find it there. It had 2 folders inside it adblock custom resources and sessions (sessions had 1 file called current inside it)._

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/browsing-history-no-longer-working/647916

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/after-windows-update-passwords-are-gone/647990

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/password-manager-empty-after-brave-update-login-data-exists-but-not-decrypting-windows/648402/4

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/can-i-recover-all-lost-data-settings-part-2/648529

**Comment by Saoiray** [comment]
https://community.brave.app/t/after-the-latest-update-the-browser-is-failing-to-recognize-or-load-existing-user-profile-data-passwords-history-bookmarks-and-settings-even-though-the-physical-data-files-remain-intact-in-the-local-directory/648539

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/how-to-restore-crashed-session/649026/2

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-update-wiped-all-passwords-history-and-settings-no-warning-locked-out-of-accounts/651037

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/lost-all-the-saved-passwords-after-brave-update/651235

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-update-wiped-all-passwords-history-and-settings-no-warning-locked-out-of-accounts/651037/18?u=mattches

**Comment by Brave-Matt** [comment]
https://community.brave.app/t/brave-removed-all-of-my-saved-password-bookmarks-and-logged-me-out-of-all-sites/651602/2

**Comment by rubgithub** [comment]
This is a SH#$%$ my pass seed doesn't work

**Comment by Isaac-Clarke** [comment]
I have changed CPU, MOBO and RAM these last days.
Today I log in, get into Brave, and see that NOT ONLY I am log out of every account, but my whole passwords list has been **purged!**
_**VERY IMPORTANT TO NOTICE!!! :** I did NOT get ANY popup or ANYTHING! The passwords were just gone (yesterday they weren't, when the new pieces were already in, while today they are!). [Here's some extra info](https://discussion.fedoraproject.org/t/my-fedora-system-often-hangs-up-crashes-on-shutdown/139951/28)._

Now I am a lucky autistic man, so I have some backuped password exports, _so the ONLY thing which is continuing to happen is a slight sense of anxiety_, but if I didn't have them and the passwords were just gone **I'd've needed to go to the flipping hospital by the sheer gravity of the loss!!!** (Normal people don't use password managers and don't expect their servant-software to randomly poop the bed.)

It'd be nice to be able to get the most recent version of all passwords back (I don't remember if I made new accounts or if I changed some passwords), but **MOST IMPORTANT OF ALL it is an ABSOLUTE IMPERATIVE to make this NEVER happen EVER again, because it's a failure on the software developers' side.**

(You are free to read this as an angry drunk at a bar, but I don't intend to come through like that, imagine more the Mr Incredible meme going "Math is MATH!"... I just hope the mods don't see this message as an "attack" on the Devs or whoever, because they'd just be silencing what polite software users are **actually** thinking about when these things happen... As said above, I am LUCKY I don't have to go to the hospital, and YES, I am NOT exagerating, and NO, I am not doxxing myself to "prove" it. Being VERY angry that something like this happen is NORMAL HUMAN BEHAVIOR. "Very angry" is the bare minimum anyone and everyone **without a password backup** WILL feel!)

It's one thing if my laptop/desktop suffers a power surge, gets crushed by a truck, or has a cosmic ray fry the SSD, but it's another that something as simple as _"oopsies, the key is corrupted! teh eh!"_ **DESTROYS** all my damn passwords!
(**Separate BUT related** issue is that [you just need a power surge to be completely locked out of Brave!](https://community.brave.app/t/how-to-unlock-profile-that-keeps-brave-from-starting/526455) I had to delete the USELESS "key / lock" files **on ALL of my PCs** to avoid this ISSUE, PROBLEM, BUG, USELESS ANTI-USER FEATURE!)

**Comment by evanrgardner** [comment]
+1 a couple new reports provided via https://community.brave.app/t/notice-unexplained-data-loss-issue/485498/194?u=evan123

**Comment by Isaac-Clarke** [comment]
Small update on my side:

It turned out that my RAM had a malfunctioning stick.

Altho the malfunction of the stick is provable, objective (check [this Fedora Forum page](https://discussion.fedoraproject.org/t/system-fedora-kde-lagged-froze-and-needed-hard-restart-many-times-now-its-worse/200196/38) for more info), this issue could still have been due to a failure of the software.

## tl;dr
EVEN IF my case had the hardware at fault, it doesn't make any other case less real IF they are unrelated to hardware issues.

**Comment by Isaac-Clarke** [comment]
It happened again, with a fresh install of Fedora and completely functioning hardware.

<img width="480" height="640" alt="Image" src="https://github.com/user-attachments/assets/dbc015bc-b205-4125-a5d9-55f25b347642" />

Is there, like, a [stupid key system](https://community.brave.app/t/how-to-unlock-profile-that-keeps-brave-from-starting/526455) for passwords too, or something? Because this is
# UNACCEPTABLE.

---

NORMAL people don't keep backups of ANYTHING. So where user error is to be blamed on **no one but the user,** this, is NOT, "user error".


I basically got these two [on my main PC too](https://discussion.fedoraproject.org/t/indescribable-desktop-enviroment-dies-and-everything-unmounts-but-the-cursor-persists/201323/19):

<img width="779" height="435" alt="Image" src="https://github.com/user-attachments/assets/f27294c2-f7ad-449a-8349-c1f13a4e5c78" />

<img width="779" height="338" alt="Image" src="https://github.com/user-attachments/assets/6d36fbd2-b989-4ed0-bfe9-b30e0642e7f2" />

Then I used the PC, an insect decided to dirty my table so I shut down the PC **by doing** "shut down", then closing Brave, then confirming, then the PC shut itself off (the "confirm me" popup **allows Brave to deny** the PC's ability to shut itself off normally) and now that I'm back the passwords were gone again! :)

---

I have NOTHING else to say!
I can't say more to describe the issue further BECAUSE there's nothing more I can describe!
This is ABSOLUTELY a weakness of Brave and it MUST have been fixed since it came to be!!!

**Comment by itspmohit** [comment]
I have noticed that if I close the browser using `pkill -TERM brave`, All the session cookies, etc are lost and I am logged out of all the websites. This can also consistently reproduced by keeping the browser open and logging out of the session or rebooting or shutting down the computer. Upon reopening the browser, All sessions/cookies are deleted and I am logged out of all the websites.
On the other hand If I close the browser (Using Alt+F4 or Mod+Shift+Q) and reopen it then the session/cookies are not lost and I am NOT logged out of websites.
I am using Asahi-Fedora with Sway. I have also noticed this issues on Asahi-Fedora Gnome/KDE. I am running it on M1 Pro Macbook Pro.

I have been using Brave on MacOS before but I never faced this issue.

**Comment by itspmohit** [comment]
I can consistently reproduce this on Linux when Brave is not closed through its normal window/UI shutdown path.

### Steps to reproduce

1. Open Brave and log into several websites.
2. Leave Brave running.
3. Exit the desktop session by logging out, rebooting, or shutting down the system **without first closing Brave**.
4. Start the system and reopen Brave.

### Actual result

All website session cookies appear to be lost, and I am logged out of all websites.

The same behavior can also be reproduced without rebooting by terminating Brave externally:

```bash
pkill -TERM -x brave
```

and also with:

```text
brave://quit
```

In both cases, reopening Brave results in all website sessions being lost.

### Comparison

If I close Brave normally using `Alt+F4` or my Sway keybinding (`Mod+Shift+Q`) and then reopen it, the cookies/session state is preserved and I remain logged in.

Therefore, the issue appears to be specific to how Brave persists/restores cookie/session state during certain shutdown/quit paths rather than simply being caused by the desktop session being terminated.

I initially suspected the Linux Secret Service/keyring, but Brave has access to GNOME Keyring and has created the `Brave Safe Storage` entry. The behavior is also independent of the desktop environment: I can reproduce it on Asahi Fedora with Sway, and I have also observed it on Asahi Fedora with GNOME and KDE.

Hardware: 16" MacBook Pro M1 Pro running Asahi Fedora.

I previously used Brave on macOS and did not encounter this issue there.

**Comment by itspmohit** [comment]
I tested it with the latest version of Chromium and this issue is **not** present there.

```text
Chromium	152.0.7977.82 (Official Build) Built from source for Fedora release 44 (Forty Four) (arm64) 
Revision	d04cdb24d67b081f6cf80200ffc5233f44b61109-refs/branch-heads/7977@{#2322}
OS	Linux
JavaScript	V8 15.2.124.21
User Agent	Mozilla/5.0 (X11; Linux aarch64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/152.0.0.0 Safari/537.36
Command Line	/usr/bin/chromium-browser --enable-plugins --enable-extensions --enable-user-scripts --enable-printing --enable-sync --auto-ssl-client-auth --ozone-platform=wayland --flag-switches-begin --flag-switches-end --render-node-override=/dev/dri/renderD128
Executable Path	/usr/lib64/chromium-browser/chromium-browser.sh
```
I installed Chromium and changed its settings to match what I use on Brave.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/bf45ebf7-08a5-4370-b57e-3c6225503346" />

I also use "Delete Data on Exit" setting in Brave which is not present in Chromium.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/d29284c7-04c4-4c02-9466-5bc1a6d20588" />

I started the browser, Then changed some settings on google search and checked if cookies are allowed from the site. Then I quit using `chrome://quit` and `pkill -TERM chromium-browse`. In both cases the cookies presisted and the settings for google.com were not lost like in case of Brave.

Since the "Delete data on exit" setting is the only difference between Brave and Chromium, I disabled it on Brave and then tried to reproduce the issue with data loss that I was having. And, I found that. Disabling the "Delete data on Exit" **Fixed** the issue of cookies/session lost when Brave is not closed through its normal window/UI shutdown path for me.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/496144c7-bc66-4cd2-aabe-e0ff2e1449b2" />

This bug seems to be in the implementation of "Delete data on exit" setting in Brave. I never checked "Cookies and other site data" in "Delete data on exit" setting and this setting was respected when closing Brave normally. But this setting was ignored and the data of all sites is purged when not closing the browser normally.

**Comment by Isaac-Clarke** [comment]
> Since the "Delete data on exit" setting is the only difference between Brave and Chromium, I disabled it on Brave and then tried to reproduce the issue with data loss that I was having. And, I found that. Disabling the "Delete data on Exit" **Fixed** the issue of cookies/session lost when Brave is not closed through its normal window/UI shutdown path for me.


> This bug seems to be in the implementation of "Delete data on exit" setting in Brave. I never checked "Cookies and other site data" in "Delete data on exit" setting and this setting was respected when closing Brave normally. But this setting was ignored and the data of all sites is purged when not closing the browser normally.

It's off by default. How even do you "disable" it?

**Comment by itspmohit** [comment]
> It's off by default. How even do you "disable" it?

Yes, It off by default. But I keep it on to clear the history and cache when exiting the browser so I don't have to manually do it.
I have it turned on for as long as I remember. And I was facing the cookies/session purge issue on Linux since last month (I was daily driving macOS previously and don't remember facing this issue there).
So I unchecked all the boxes like it is by default and that seemed to fix the issue I was facing.

**Comment by Isaac-Clarke** [comment]
> Yes, It off by default.

Then you didn't fix it.
You can go back to all the details I've already wrote **and won't repeat again** because it's **a lot.**
Your situation is different.

**Comment by itspmohit** [comment]
> > Yes, It off by default.
> 
> Then you didn't fix it. You can go back to all the details I've already wrote **and won't repeat again** because it's **a lot.** Your situation is different.

It seems like I have found a new bug. I'll file a new bug report for this.

**Comment by bsclifton** [comment]
Hi folks - I wanted to share an update

This is a very complex problem. Thanks for all of your patience. I've been tracking this problem for a long time. I've made various hypothesis, tested them, etc. The notes that everybody has shared both here and on Community are wonderful.

To be clear - this is a bug in Chromium which we share. 
But I have made some observations and I believe we can make things better in Brave. If those prove to be successful, I would be happy to upstream the work to Chromium also.

**_Here are two problems which can happen which cause folks to lose data:_**

## Person can have a corrupt `Local State` file. 
`Local State` is a file at the profile root (ex: `%LOCALAPPDATA%\BraveSoftware\Brave-Browser\User Data\Local State`) and it is a very important file. There is a section inside of there called `os_crypt` which has an encrypted key. This key (when decrypted) is used to save passwords, cookies, and more.  If the `Local State` file has ever gotten corrupted, you would have a `Local State.bad` file (ex: `%LOCALAPPDATA%\BraveSoftware\Brave-Browser\User Data\Local State.bad`). 

The `Local State` file itself is JSON and there are a few ways it can become corrupt. 
- For the sake of example, someone can reproduce this by closing browser, opening their `Local State` in Notepad, typing a character in the middle of the file (ex: put a random `}`), and saving it / relaunch browser. 
- In the real world, this type of corruption can happen when the `Local State` is being saved. If you open `Local State` in an editor like SublimeText and keep the browser open, you'll get prompted to reload the file every ~30 seconds. This is because the browser is saving the file periodically.  Someone can kill the browser in the middle of one of these saves (ex: if you click shutdown, it says Brave is still running, you click "end now"). **I suspect the `Local State` file writing is not atomic and the browser gets killed mid-stream, in the middle of writing to this file**. Next launch would trigger the corruption. 

In either case, a `Local State.bad` file fill be created and the person will be shown a dialog message saying there was a problem. Brave and other Chromium browsers do not recover from this.

### Options
1. We can trap the situation where `Local State` is corrupt and try to load the old OSCrypt key from `Local State.bad`.
2. We could keep a history for the OSCrypt key. This key is encrypted - it's safe to store a copy. We could even have a history.

## Person can have their OS encryption key changed
My findings here are specific to Windows - but I do want to share them. There are two kinds of encryption that can be used for OSCrypt: DPAPI and App Bound Encryption. App Bound Encryption is only used for situations where there is a system level install - I need to look at this more closely.

For DPAPI, I have verified there are situations where this key is re-issued. In most cases, the OS should have a history of the key. There can be a policy in place to expire DPAPI key every X number of days also. In both of these cases, we should recover fine and the person wouldn't see an impact.  However, there are scenarios where the person could be broken. One example I am aware of is: the person is joined to a domain and they have their account password reset by an administrator. **This will cause the OS to trash the DPAPI password and there is no way to recover it.**  

Above, I described the `Local State` file. The `os_crypt` section has a key section (the Brave key). But it's important to note: that key is an encrypted key. And OSCrypt is using DPAPI to encrypt/decrypt it. If the OS loses the key, you can't decrypt the Brave key. If you can't decrypt the Brave key, you lose your passwords and cookies.  This is why copying your profile from one machine to another will cause you to lose passwords and cookies for cases where OSCrypt is used.

### Options
This is somewhat preventable. A person could use Brave Sync and they will use another encryption mechanism altogether. This is ultimately what I would recommend. Chrome seems to push folks towards the Google sign-in and I suspect that is largely how Chrome avoids this problem. If they're using sync, they can't lose the secret.

I have some heuristics for detecting. If we detect they are in those conditions, we can then try to guide folks towards sync. I should be able to put together a help article with all the details and the UI would link to that.

## Conclusion
Thanks again for your patience. I have some code in a branch - I'm looking at this. There's been a lot of activity lately. I wish this was an easier problem to solve. But we need to find the root cause and the correct solution.

I'll go ahead and self-assign this issue as I'll have a patch relatively soon for the items discussed above.

**Comment by Isaac-Clarke** [comment]
Excuse me @bsclifton , I am not a code guy so I am not even gonna start engaging on that, but isn't it easier, at least for a quick fix, to "keep two copies of that file" as a "backup"?

The error comes to be because the file gets corrupted in some way, because the file gets written every couple of moments, and if **one** of these writes goes wrong the file is gone.
Can't this be prevented by updating 2 distinct files asynchronously, so that when one gets corrupted you still have the other thus avoiding DISASTROUS data loss?

The browsing history isn't THAT important, BUT the passwords list is!

I'd advice to do that instead, IMMEDIATELY, while the tech boys in the back fix the problem properly.
**The PRIORITY is to avoid ANY AND ALL further and future PASSWORDS loss!**

**Comment by deyama2012** [comment]
Just wanted to chime in with a security consideration: moving user data from one machine to another can also be malicious, e.g. as part of an infostealing attack. In such a scenario, I'd prefer that the data be lost rather than safely transferred to the new machine.

I don't know if there is a practical way to reliably distinguish a legitimate migration from a malicious one, but I think this is worth keeping in mind when implementing the feature.

**Comment by bsclifton** [comment]
@Isaac-Clarke - I'm working through a solution to do a backup. It's in the following pull request:
https://github.com/brave/brave-core/pull/40058

There are some specifics I need to fix up before we can land it - but it's a good improvement.

@deyama2012 copying from other machines - you won't be able to decrypt that. There's a Brave key and that is encrypted with the OS key. The only way another device will be able to decrypt is if you use sync. You're expected to lose your encrypted data if you copy the profile from one machine to the other and you are not using sync.
